File: leetcode/python/310_minimum_height_trees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: list[list[int]]) -> list[int]:
        if n <= 2:
            return [i for i in range(n)]

        neighbors = self._transformEdges(edges)
        return self._findCentroids(neighbors)
        
    def _transformEdges(self, edges: list[list[int]]) -> list[set[int]]:
        neighbors = [set[int]() for _ in range(len(edges) + 1)]
        for [start, end] in edges:
            neighbors[start].add(end)
            neighbors[end].add(start)
        return neighbors

    def _findCentroids(self, edges: list[set[int]]) -> list[int]:
        leaves: list[int] = []
        for (i, nodes) in enumerate(edges):
            if len(nodes) == 1:
                leaves.append(i)

        remaining = len(edges)
        while remaining > 2:
            remaining -= len(leaves)
            tmp: list[int] = []
            for leaf in leaves:
                parent = edges[leaf].pop()
                edges[parent].remove(leaf)
                if len(edges[parent]) == 1:
                    tmp.append(parent)
            leaves = tmp
        return leaves

File: leetcode/rust/src/310_minimum_height_trees.rs
impl Solution {
    pub fn find_min_height_trees(n: i32, edges: Vec<Vec<i32>>) -> Vec<i32> {
        if n <= 2 {
            return (0..n).collect();
        }
        let mut graph = vec![vec![]; n as usize];
        let mut degree = vec![0; n as usize];
        for edge in edges {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            graph[u].push(v);
            graph[v].push(u);
            degree[u] += 1;
            degree[v] += 1;
        }
        let mut leaves = Vec::new();
        for i in 0..n as usize {
            if degree[i] == 1 {
                leaves.push(i);
            }
        }
        let mut remaining = n;
        while remaining > 2 {
            remaining -= leaves.len() as i32;
            let mut new_leaves = Vec::new();
            for leaf in leaves {
                for &neighbor in &graph[leaf] {
                    degree[neighbor] -= 1;
                    if degree[neighbor] == 1 {
                        new_leaves.push(neighbor);
                    }
                }
            }
            leaves = new_leaves;
        }
        leaves.into_iter().map(|x| x as i32).collect()
    }
}
